// eslint-plugin-lodash/lib/rules/chain-style.js
/**
 * @fileoverview Rule to enforce a specific chain style
 */
'use strict';

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = function (context) {
    var _require = require('../util/lodashUtil');

    var isChainable = _require.isChainable;
    var isChainBreaker = _require.isChainBreaker;
    var isExplicitChainStart = _require.isExplicitChainStart;
    var isImplicitChainStart = _require.isImplicitChainStart;

    var _require2 = require('../util/astUtil');

    var isMethodCall = _require2.isMethodCall;

    var settings = require('../util/settingsUtil').getSettings(context);
    var callExpressionVisitors = {
        'as-needed': function asNeeded(node) {
            if (isExplicitChainStart(node, settings.pragma)) {
                var curr = node.parent.parent;
                var needed = false;
                while (isMethodCall(curr) && !isChainBreaker(curr, settings.version)) {
                    if (!isChainable(curr, settings.version) && !isChainBreaker(curr.parent.parent, settings.version)) {
                        needed = true;
                    }
                    curr = curr.parent.parent;
                }
                if (isMethodCall(curr) && !needed) {
                    context.report(node, 'Unnecessary explicit chaining');
                }
            }
        },
        implicit: function implicit(node) {
            if (isExplicitChainStart(node, settings.pragma)) {
                context.report(node, 'Do not use explicit chaining');
            }
        },
        explicit: function explicit(node) {
            if (isImplicitChainStart(node, settings.pragma)) {
                context.report(node, 'Do not use implicit chaining');
            }
        }
    };

    return {
        CallExpression: callExpressionVisitors[context.options[0] || 'as-needed']
    };
};

module.exports.schema = [{
    enum: ['as-needed', 'implicit', 'explicit']
}];
