// eslint-plugin-lodash/lib/rules/prefer-constant.js
/**
 * @fileoverview Rule to check if the expression could be better expressed as a _.constant
 */
'use strict';

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = function (context) {
    var _require = require('../util/astUtil');

    var getValueReturnedInFirstLine = _require.getValueReturnedInFirstLine;

    var shouldCheckArrowFunctions = context.options[0] !== undefined ? context.options[0] : true;
    var shouldCheckFunctionDeclarations = context.options[1] !== undefined ? context.options[1] : false;

    function isCompletelyLiteral(node) {
        switch (node.type) {
            case 'Literal':
                return true;
            case 'BinaryExpression':
                return isCompletelyLiteral(node.left) && isCompletelyLiteral(node.right);
            case 'UnaryExpression':
                return isCompletelyLiteral(node.argument);
            case 'ConditionalExpression':
                return isCompletelyLiteral(node.test) && isCompletelyLiteral(node.consequent) && isCompletelyLiteral(node.alternate);
            default:
                return false;
        }
    }

    function reportIfLikeConstant(func, node) {
        var valueReturnedInFirstLine = func(node);
        if (valueReturnedInFirstLine && isCompletelyLiteral(valueReturnedInFirstLine)) {
            context.report(node, 'Prefer _.constant over a function returning a literal');
        }
    }

    function handleFunctionDefinition(node) {
        reportIfLikeConstant(getValueReturnedInFirstLine, node);
    }

    return {
        FunctionExpression: handleFunctionDefinition,
        FunctionDeclaration: function FunctionDeclaration(node) {
            if (shouldCheckFunctionDeclarations) {
                handleFunctionDefinition(node);
            }
        },
        ArrowFunctionExpression: function ArrowFunctionExpression(node) {
            if (shouldCheckArrowFunctions) {
                handleFunctionDefinition(node);
            }
        }
    };
};

module.exports.schema = [{
    type: 'boolean'
}, {
    type: 'boolean'
}];
